Integrate time param - term type switch

mpowelson opened the pull request:
Time param sections are still skeletons. But it doesn't appear to have broken anything without time.

Levi-Armstrong commented:
Ready to be rebased.

mpowelson commented:
@Levi-Armstrong Just rebased. Also, this would address #13 .

Levi-Armstrong commented on trajopt/include/trajopt/problem_description.hpp:
```diff
@@ -87,6 +99,10 @@ struct BasicInfo
   sco::ModelType convex_solver; // which convex solver to use
+
+  // optional, but must be valid
+  double dt_upper_lim = 1.0;
```
Missing Doxygen

Levi-Armstrong commented on trajopt/include/trajopt/problem_description.hpp:
```diff
@@ -60,6 +70,8 @@ class TRAJOPT_API TrajOptProb : public sco::OptProb
   TrajArray GetInitTraj() { return m_init_traj; }
   friend TrajOptProbPtr ConstructProblem(const ProblemConstructionInfo&);
+  // Indicates whether or not the last column is dt
+  bool has_time;
```
Recommend moving this to private and providing a Get and Set method.

Levi-Armstrong replied:
Also add doxygen.

Levi-Armstrong commented on trajopt/include/trajopt/problem_description.hpp:
```diff
@@ -97,10 +113,13 @@ struct InitInfo
+  double dt = 1.0;
 };
 
 struct TRAJOPT_API MakesCost
```
Should these be removed?

Levi-Armstrong commented on trajopt/src/problem_description.cpp:
```diff
@@ -208,16 +234,17 @@ void ProblemConstructionInfo::readInitInfo(const Json::Value& v)
-  Eigen::VectorXd start_pos = env->getCurrentJointValues(kin->getName());
 
-  if (type_str == "stationary")
+  if (boost::iequals(type_str, "stationary"))
```
Should we use `type_str.compare("stationary") == 0` instead of boost?

mpowelson replied:
This is something Reid had in there so that it would be case insensitive. I'll go with whatever you recommend if there's a better way of doing this (or if you'd prefer it to be case sensitive)

mpowelson commented:
@Levi-Armstrong Made those changes. use_time is now in basic_info, and added a check in construct problem to catch when it is not set when using the C++ interface. 

I also added a unit test that checks that the initial trajectory is the right size, and also (while I was at it) that the joint_interpolated (formally straight_line) initialization is working as expected.

mpowelson commented:
@Levi-Armstrong I believe I have added all of things we discussed this morning.

mpowelson commented:
Also, do we need to update the plotting in Tesseract somehow? I had a branch that added a -1 in ros_basic_plotting.h, but that code has been removed now.

Levi-Armstrong commented:
@mpowelson It looks like when I merged the PR to fix warning it created a few conflicts

mpowelson commented:
@Levi-Armstrong Rebased. It should be fine now.

Levi-Armstrong commented on trajopt/include/trajopt/problem_description.hpp:
```diff
@@ -50,6 +51,14 @@ class TRAJOPT_API TrajOptProb : public sco::OptProb
   TrajOptProb();
   TrajOptProb(int n_steps, const ProblemConstructionInfo& pci);
   virtual ~TrajOptProb() = default;
+  sco::VarVector GetJointVarRow(int i)
```
I recommend using the `rblock` function in BasicArray then change this method to GetVarRow(int i, int start_col, int num_col).

mpowelson replied:
This has been changed.

Levi-Armstrong commented:
@mpowelson A few minor request then it is ready to be merged

mpowelson commented on trajopt/include/trajopt/problem_description.hpp:
```diff
@@ -133,8 +173,12 @@ struct TRAJOPT_API TermInfo
+
 private:
   static std::map<std::string, MakerFunc> name2maker;
+  int supported_term_type_;
```
Yes it should be.

Levi-Armstrong commented:
Is this ok to merge?

mpowelson commented:
It should be.
